cluster/addons/metadata-agent/stackdriver/metadata-agent.yaml

```diff
@@ -56,6 +56,13 @@ spec:
       dnsPolicy: ClusterFirst
       restartPolicy: Always
       schedulerName: default-scheduler
+      tolerations:
+      - key: "node.alpha.kubernetes.io/ismaster"
+        effect: "NoSchedule"
+      - operator: "Exists"
+        effect: "NoExecute"
+      - operator: "Exists"
+        effect: "NoSchedule"
```

Comment:
Do we really want such strong tolerations? This will make this able to schedule anywhere, including for example a node that is currently being deleted or a super-expensive node with GPU.

Reply:
This is what we currently have for fluentd and they should really go together. If this is more restricted, then fluentd should be more restricted, too.

Reply:
Ah, so this is meant to run on every node? In this case it makes much more sense.

What about `metadata-agent-cluster-level` below though? If it's one-per-cluster than presumably it should schedule on a node without taints, no?

Reply:
Good point. Yes, I think the cluster level one should respect taints.

Reply:
We do want all cluster level Metadata Agent to be scheduled though. Otherwise we lose insight of the metadata. It's intentional to disregard the taints. Note that this will only apply to customers who enable the Metadata Agent addons.

Reply:
@x13n @MaciekPytel  WDYT?

Reply:
This is similar to Heapster - we want to always have it running. And Heapster respects NoSchedule/NoExecute taints, having a toleration for CriticalAddonsOnly. I think the cluster-level metadata agent should follow the same pattern.

Reply:
@x13n - Yeah, `Heapster` does use `CriticalAddonsOnly` 
https://github.com/kubernetes/kubernetes/blob/2e0e1681a6ca7fe795f3bd5ec8696fb14687b9aa/cluster/addons/cluster-monitoring/stackdriver/heapster-controller.yaml#L126

As I read this [guaranteed-scheduling-critical-addon-pods doc](https://kubernetes.io/docs/tasks/administer-cluster/guaranteed-scheduling-critical-addon-pods/) to understand a bit more about what it does, and it sounds like `CriticalAddonsOnly` will be deprecated soon. Instead, it recommends [priorityClass](https://kubernetes.io/docs/concepts/configuration/pod-priority-preemption/#priorityclass).

And I did find a `priorityClassName` spec defined for Heapster:
https://github.com/kubernetes/kubernetes/blob/2e0e1681a6ca7fe795f3bd5ec8696fb14687b9aa/cluster/addons/cluster-monitoring/stackdriver/heapster-controller.yaml#L48

We probably should just use `priorityClassName: system-cluster-critical` for Logging and Metadata Agents then?

Reply:
@x13n @MaciekPytel Friendly ping.

Reply:
https://kubernetes.io/docs/concepts/configuration/pod-priority-preemption/ says this feature is in beta starting from 1.11. You can add it, but don't cherrypick it to 1.10 branch later.

Reply:
Cool. Updated to use `priorityClassName: system-cluster-critical` instead. PTAL

Reply:
@x13n @MaciekPytel